lib/report.js

```diff
@@ -158,7 +160,162 @@ class ConciseReporter extends Reporter {
   }
 }
 
+class TapReporter extends Reporter {
+
+  constructor(options) {
+    super(options);
+    this.successful = 0;
+    this.failed = 0;
+    this.skipped = 0;
+    this.failedTests = [];
+    this.testNumeration = [1];
+    this.lastIndex = 0;
+    this.message = '';
+    this.firstFailure = true;
+    this.got = [];
+    this.expected = [];
+  }
+
+  error(test, error) {
+    // remove one success because 'error' may only be received after end
+    // though the test may have already failed so we need a way to check
+    // if we should do this, i.e. store map of successful tests
+    // this.successes--;
+    // test.todo ? this.skipped++ : this.failures++;
+    // TODO(lundibundi): add test.metadata.filename here too once available
+    this.listFailure(
+      test,
+      { message: `Failure after end: ${error ? error.message : 'no message'}` },
+      ` Test failed: ${test.caption}`
+    );
+  }
+
+  lastElement(inc) {
+    if (inc) this.testNumeration[this.lastIndex]++;
+    return this.testNumeration[this.lastIndex];
+  }
+
+  calculateSpaces() {
+    this.message = pad(this.lastIndex * 2);
+  }
+
+  parseTestResults(test) {
+    if (test) {
+      this.lastElement(1);
+      if (!test.success) {
+        this.failedTests.push(this.testNumeration[this.lastIndex] - 1);
+        this.listFailure(test, true);
+        this.got.forEach(el => {
+          this.message += `${pad(6)}- ${el}\n`;
+        });
+        this.message += `${pad(4)}expect: \n`;
+        this.expected.forEach(el => {
+          this.message += `${pad(6)}- ${el}\n`;
+        });
+        this.message += '  ...\n';
+        this.got = [];
+        this.expected = [];
+        console.log(this.message);
+      } else {
+        this.testNumeration.push(0);
+        this.lastIndex++;
+        this.calculateSpaces();
+        test.results.forEach(res => {
+          if (res.success) {
+            if (res.type === 'subtest') {
+              const t = res.test;
+              console.log(`${this.message}# Subtest: ${t.caption}`);
+              this.parseTestResults(t);
+              this.calculateSpaces();
+              const prefix = t.success ? 'ok' : 'not ok';
+              let message1 = `${this.message}${prefix} ${this.lastElement(0)}`;
```

Comment:
I'd prefer that it was just `message` instead of `message1` especially because there is no such variable in this scope.